thetis/timeintegrator.py

```diff
@@ -695,342 +324,96 @@ def advance(self, t, dt, solution, updateForcings=None):
+            solve(self.mass_matrix, self.solution, self.rhs_func)
 
-        # Allocate tendency fields
-        self.k = []
-        for i in xrange(self.n_stages):
-            fname = '{:}_k{:}'.format(self.name, i)
-            self.k.append(Function(fs, name=fname))
-        # construct variational problems
-        self.F = []
-        if not mixed_space:
-            for i in xrange(self.n_stages):
-                for j in xrange(i+1):
-                    if j == 0:
-                        u = self.solution_old + self.a[i][j]*self.dt_const*self.k[j]
-                    else:
-                        u += self.a[i][j]*self.dt_const*self.k[j]
-                self.F.append(-inner(self.k[i], test)*dx +
-                              self.equation.residual(terms_to_add, u, self.solution_old, fields, fields, bnd_conditions))
-        else:
-            # solution must be split before computing sum
-            # pass components to equation in a list
-            for i in xrange(self.n_stages):
-                for j in xrange(i+1):
-                    if j == 0:
-                        u = []  # list of components in the mixed space
-                        for s, k in zip(split(self.solution_old), split(self.k[j])):
-                            u.append(s + self.a[i][j]*self.dt_const*k)
-                    else:
-                        for l, k in enumerate(split(self.k[j])):
-                            u[l] += self.a[i][j]*self.dt_const*k
-                self.F.append(-inner(self.k[i], test)*dx +
-                              self.equation.residual(terms_to_add, u, self.solution_old, fields, fields, bnd_conditions))
-        self.update_solver()
+    def eval_rhs(self):
+        if self._nontrivial:
+            assemble(self.l, self.rhs_func)
 
-    def update_solver(self):
-        # construct solvers
-        self.solver = []
-        for i in xrange(self.n_stages):
-            p = NonlinearVariationalProblem(self.F[i], self.k[i])
-            sname = '{:}_stage{:}_'.format(self.name, i)
-            self.solver.append(
-                NonlinearVariationalSolver(p,
-                                           solver_parameters=self.solver_parameters,
-                                           options_prefix=sname + '_k{}'.format(i)))
+    def correct(self):
+        """
+        Correction step from t_{n} to t_{n+1}
 
-    def initialize(self, init_cond):
-        """Assigns initial conditions to all required fields."""
-        self.solution_old.assign(init_cond)
+        (MT)_{n+1} = (MT)_{n} + dt*L_{n+1/2}
 
-    def advance(self, t, dt, solution, update_forcings=None):
+        This is Euler ALE step: LHS is evaluated in Omega_{n+1}, RHS in Omega_n
+        """
+        if self._nontrivial:
+            # NOTE must call eval_rhs in the old mesh first
+            self.rhs_func += self.msolution_old
+            assemble(self.a, self.mass_matrix)
+            solve(self.mass_matrix, self.solution, self.rhs_func)
```

Comment:
If the mass matrix is DG, I think you're better off (performance-wise) directly assembling the inverse:

```python
assemble(self.a, self.mass_matrix, inverse=True)
with self.solution.dat.vec as x:
    with self.rhs_func.dat.vec_ro as b:
        self.mass_matrix.force_evaluation()
        self.mass_matrix.petscmat.mult(b, x)
```

The same in the `predict` step too.

It's not obvious why the mass_matrix changes between steps (I guess due to outside ALE?  Maybe mention that?)